File: Homework-1/an_nguyen_task2.py
import sys
import json
import time
from pyspark.sql import SparkSession


def main(input_file, output_file, num_of_partitions):
    result = {}
    spark = SparkSession.builder.master("local[*]").appName("HW1").getOrCreate()
    text = spark.sparkContext.textFile(input_file).cache()

    """ Default partition """
    start = time.time()
    default = text.coalesce(8)
    temp = default                                          \
        .map(json.loads)                                    \
        .map(lambda x: [x["user_id"], x["review_count"]])   \
        .map(lambda x: [x[0], x[1]])                        \
        .sortBy(lambda x: (-x[1], x[0]))                    \
        .take(10)
    size = default.glom().map(len).collect()
    process_time = time.time() - start
    result["default"] = {"n_partition": default.getNumPartitions(), "n_items": size, "exe_time": process_time}

    """ Customized partition """
    start = time.time()
    num_of_partitions = int(num_of_partitions)
    customized = text.repartition(num_of_partitions)
    temp = customized                                       \
        .map(json.loads)                                    \
        .map(lambda x: [x["user_id"], x["review_count"]])   \
        .map(lambda x: [x[0], x[1]])                        \
        .sortBy(lambda x: (-x[1], x[0]))                    \
        .take(10)
    size = customized.glom().map(len).collect()
    process_time = time.time() - start
    result["customized"] = {"n_partition": num_of_partitions, "n_items": size, "exe_time": process_time}

    """ Explanation """
    result["explanation"] = "In Default paritioning (including using coalesce), the size of each chunks is vastly" + \
                            " different while they are more uniformed when created by Customize partitioning" + \
                            " (using repartition). This is because rapartition will reshuffle and divide the data" + \
                            " equally while Default partitioning will just assign chunks of data based on their" + \
                            " location."

    with open(output_file, "w") as f:
        json.dump(result, f)


if __name__ == "__main__":
    main(sys.argv[1], sys.argv[2], sys.argv[3])

File: Homework-1/an_nguyen_task2.scala
import java.io.{File, PrintWriter}
import org.json4s._
import org.json4s.jackson.JsonMethods._
import org.apache.log4j.{Level, Logger}
import org.apache.spark.sql.SparkSession


object an_nguyen_task2 {
	implicit val formats: DefaultFormats.type = DefaultFormats

	case class User(user_id: String, name: String, review_count: Int, yelping_since: String, friends: Any,
					useful: Int, funny: Int, cool: Int, fans: Int, elite: Any, average_stars: Float,
					compliment_hot: Int, compliment_more: Int, compliment_profile: Int, compliment_cute: Int,
					compliment_list: Int, compliment_note: Int, compliment_plain: Int, compliment_cool: Int,
					compliment_funny: Int, compliment_writer: Int, compliment_photos: Int
				   )

	def main(args: Array[String]): Unit = {
		Logger.getLogger("org").setLevel(Level.INFO)
		val spark = SparkSession
			.builder()
			.appName("an_nguyen_task1")
			.config("spark.master", "local[*]")
			.config("spark.executor.memory", "6g")
			.getOrCreate()
		val text = spark.sparkContext.textFile(args(0))

		val pw = new PrintWriter(new File(args(1)))
		pw.write("{")

		// Default Partition
		pw.write("\"default\": {")
		var start = System.currentTimeMillis()
		val default = text.coalesce(8)
		var temp = default
			.map(convertToJson)
			.map(x => (x.user_id, x.review_count))
			.map(x => (x._1, x._2)).sortBy(x => (-x._2, x._1), ascending = false).take(10)
		var size = default.glom().map(_.length).collect()
		var processTime = (System.currentTimeMillis() - start).toFloat / 1000.toFloat
		pw.write("\"n_partition\":" + default.getNumPartitions + ",\"n_items\":[")
		for (i <- size) {
			pw.write(i)
			if (size.indexOf(i) == size.length) {
				pw.write(",")
			}
		}
		pw.write("]")
		pw.write(",\"exe_time\":" + processTime)
		pw.write("},")

		// Customize Partition
		pw.write("\"customize\": {")
		start = System.currentTimeMillis()
		val customized = text.repartition(args(2).toInt).coalesce(8)
		temp = customized
			.map(convertToJson)
			.map(x => (x.user_id, x.review_count))
			.map(x => (x._1, x._2)).sortBy(x => (-x._2, x._1)).take(10)
		size = customized.glom().map(_.length).collect()
		processTime = (System.currentTimeMillis() - start).toFloat / 1000.toFloat
		pw.write("\"n_partition\":" + customized.getNumPartitions + ",\"n_items\":[")
		for (i <- size) {
			pw.write(i)
			if (size.indexOf(i) == size.length) {
				pw.write(",")
			}
		}
		pw.write("]")
		pw.write(",\"exe_time\":" + processTime)
		pw.write("},")

		pw.write("\"explanation\": \"As we can see, the size of each chunks in the Customized Partitioning is much " +
			"more uniformed comparing to its Default counterpart. That is because with Default Partitioning, data " +
			"is divided based on their location while in Customized, the data is divided equally.\"")

		pw.write("}")
		pw.close()
	}

	def convertToJson(line: Any): User = {
		val temp = line + ""
		parse(temp).extract[User]
	}
}
